feat(datepicker): timeless-json-mode

davious opened the pull request:
datepickerConfig option timelessJsonMode for json models when we don't care about the time.
* also available as attribute or via datepicker-options
* on init and parse: sets model to 'yyyy-MM-dd' format
* on format: converts model to date before normal formatting
* timelessJsonUtils constant with method jsonDateToTicks for use with min-date, max-date
	* alias jsonDateAddDays
	* set timelessJsonMode = 'withUtils' to add methods to parent scope
	* also: jsonToDate -- used internally

[plunker](http://plnkr.co/edit/WzoImdpZtQcQDfN7AN5e?p=preview)
[codepen](http://codepen.io/davious/pen/qOXNZY?editors=101)

Related: #1891, #2906, #2952

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/angular-ui/bootstrap/4218)
<!-- Reviewable:end -->

wesleycho commented on src/datepicker/datepicker.js:
```diff
@@ -649,7 +663,7 @@ function($compile, $parse, $document, $rootScope, $position, dateFilter, datePar
             return undefined;
           } else {
-            return date;
+            return timelessJsonMode ? dateFilter(date, 'yyyy-MM-dd') : date;
```
This looks like it overrides the date format - shouldn't it be `dateFilter(date, dateFormat)`? What am I missing here?

davious replied:
@wesleycho this parseDate function is added to the ngModel.$parsers for non-Html5DatrInputs; see line 699; I am setting the model, usually a date object, to a date string. I want the model to be a json compatable string so I can post it over ajax without any fuss -- and I also want to have it play nicely as a date in angular-land. The 'yyyy-MM-dd' format allows this to happen.

wesleycho commented on src/datepicker/datepicker.js:
```diff
@@ -535,6 +536,19 @@ function($compile, $parse, $document, $rootScope, $position, dateFilter, datePar
+          var update = date.indexOf('T') > -1;
+          date = date.substring(0, 10).split('-');
+          date = new Date(+date[0], +date[1] - 1, +date[2]);
+          if(update) {
```
Add space after `if`

wesleycho commented on src/datepicker/datepicker.js:
```diff
@@ -684,11 +698,17 @@ function($compile, $parse, $document, $rootScope, $position, dateFilter, datePar
         ngModel.$validators.date = validator;
         ngModel.$parsers.unshift(parseDate);
         ngModel.$formatters.push(function(value) {
+          if(timelessJsonMode) {
```
Add space after `if`

wesleycho commented on src/datepicker/datepicker.js:
```diff
@@ -684,11 +698,17 @@ function($compile, $parse, $document, $rootScope, $position, dateFilter, datePar
         });
       } else {
         ngModel.$formatters.push(function(value) {
+          if(timelessJsonMode) {
```
Add space after `if`

wesleycho commented:
I need to probe this a little more, but this looks like mostly solid work, minus that question I have.

wesleycho commented on src/datepicker/datepicker.js:
```diff
@@ -535,6 +536,30 @@ function($compile, $parse, $document, $rootScope, $position, dateFilter, datePar
+          return date;
+        };
+        
+        scope.$parent.jsonToTicks = function(date) {
```
This has me concerned - this would appear to pollute the parent $scope, which should only be read for config values.

davious replied:
@wesleycho Yeah, that was nagging me too. I plan to fix this. When I go `scope.prop = function` it isn't available on the element scope. Do you happen to have any advice here: any special way of adding to this scope in a way that persists past the link function?

wesleycho replied:
Maybe provide an optional binding on the isolate scope where the user can pull the value from the directive?

davious replied:
@wesleycho OK. I understand what's happening now.

The `datepickerJsonUtils` directive that I was using in the [original commit](https://github.com/davious/angular-ui-bootstrap/commit/145578efd7a9e4ce8278f6b249b948e8a98a2fff#diff-6240fc17e068eaeef7095937a1d63eaeR842) didn't have an isolated scope, so it was actually doing the same thing as this [latest commit](https://github.com/davious/angular-ui-bootstrap/commit/e30d284aeab91ed21bfd537837cf247032b6d9e5). The only difference is the earlier way required additional markup.

Now that I realize this, I'm personally OK with the function and property living on the parent scope.

An incidental thought I had is I could move the `oneDay` property into an optional function parameter, for example: `jsonToTicks(someJsonDate, -1)`

wesleycho commented on src/datepicker/datepicker.js:
```diff
@@ -535,6 +536,30 @@ function($compile, $parse, $document, $rootScope, $position, dateFilter, datePar
+          }
+          return date.getTime ? date.getTime() : null;
+        };
+        scope.$parent.oneDay = 1000 * 60 * 60 * 24;
```
Ditto

davious commented:
@wesleycho I moved the `jsonDateToTicks` function into a `timelessDateUtils` service with a `jsonDateAddDays` alias. This allows the client to bring in the function as they see fit.

I also added a `timelessJsonUtils` convenience option to allow the client to intentionally add the service methods to the parent scope.

davious commented:
changed `timelessJsonUtils` option to `timelessJsonMode='withUtils'`

BobbieBarker commented:
LGTM, Docs, Demo, and readme need updated?

davious commented:
added `timelessJsonMode` support for `uib-datepicker`.
plunker demo: http://plnkr.co/edit/WzoImdpZtQcQDfN7AN5e?p=preview
codepin: http://codepen.io/davious/pen/qOXNZY?editors=101

had trouble with datepicker initial formatting to 'yyyy-MM-dd', but it's working in codepin even after removing the conjoined datepicker-popup

wesleycho commented on src/datepicker/datepicker.js:
```diff
@@ -33,6 +35,13 @@ angular.module('ui.bootstrap.datepicker', ['ui.bootstrap.dateparser', 'ui.bootst
   });
 
+  self.timelessJsonMode = angular.isDefined($attrs.timelessJsonMode) ?
+     $attrs.timelessJsonMode.toLowerCase() != 'false' && ($attrs.timelessJsonMode || true)
```
`($attrs.timelessJsonMode || true)` is redundant - omit this condition completely.

This whole line can be reduced to `$attrs.timelessJsonMode !== 'false'`.

davious replied:
@wesleycho `&& ($attrs.timelessJsonMode || true)` allows the `withUtils` option to be passed in.

wesleycho replied:
Ah, I see - in any case, this can still be simplified to `self-timelessJsonMode = angular.isDefined($attrs.timelessJsonMode) ? $attrs.timelessJsonMode !== 'false' && $attrs.timelessJsonMode`

I don't think it should default to true, this should be opt-in.

davious replied:
It is opt-in. The `true` is there if the attribute is added without a value. It allows for `<div><input ng-model="date" uib-datepicker-popup="MM/dd/yyyy" timeless-json-mode></div>` to trigger the mode.

wesleycho commented on src/datepicker/datepicker.js:
```diff
@@ -859,4 +902,45 @@ function(scope, element, attrs, $compile, $parse, $document, $rootScope, $positi
   };
-});
+})
+.constant('timelessJsonUtils', (function() {
```
This should not be a constant, but a service.

paulflo150 commented:
Out of curiosity, what would be the purpose of having a "timeless" option? I would think this just adds to the complexity of the datepicker without a gain - most people want a way to set the date/ time pickers to UTC without any hacks - and that would yield a reliable date only part, as well as time.

davious commented:
this dateInterceptProvider is working for me on 0.13.4 and 0.14.3
```
(function(window, angular) {
  'use strict';
  
  angular.module('app')
    .provider('dateInterceptor', dateInterceptorProvider);
    
  function dateInterceptorProvider() {
    this.$get = function() {
      return {
        response: response,
        request: request
      };
    };
    
    function response(response) {
      convertJsonToDates(response.data);
      return response;
    }
    
    function convertJsonToDates(obj) {
      for (var key in obj) {
        var value = obj[key];
        var type = typeof value;
        if (type === 'string') {
          if(/^\d{4}-\d{2}-\d{2}T/.test(value)) {
            var dateArray = value.split('T')[0].split('-');
            var date = new Date(+dateArray[0], +dateArray[1] - 1, +dateArray[2], 0, 0, 0, 0);
            date.setFullYear(+dateArray[0]);
            obj[key] = date;
          }
        } else if (type === 'object') {
          convertJsonToDates(value);
        }
      }
    }
  
    function request(config) {
      convertDatesToJson(config.data);
      return config;
    }
    
    function pad(num) {
      return (num < 10 ? '0' : '') + num;
    }
    
    function convertDatesToJson(obj) {
      if (typeof obj !== "object") {
        return;
      }

      for (var key in obj) {
        if (!obj.hasOwnProperty(key)) {
          continue;
        }
        var value = obj[key];
        if (value && angular.isDate(value) && !isNaN(value)) {
          obj[key] = value.getFullYear() + '-' + pad(value.getMonth() + 1) + - + pad(value.getDate());
        } else if (typeof value === "object") {
          convertDatesToJson(value);
        }
      }
    }
  }
})(window, window.angular);
```

davious commented:
One nit I have with the above approach (https://github.com/angular-ui/bootstrap/pull/4218#issuecomment-158249540) is it produces a `flash-of-invalid` as the date model is transformed to a string on the request.

wesleycho commented:
Question - is this PR relevant anymore after the timezone work done now in the library?

davious commented:
For the envisioned use-case, it is possible to use `ng-model-options = {timezone: utc}`; but, I still need to convert my date strings and all my client-side dates to utc dates.

From my point of view, this does not match the intent of the pr, which is to provide a no-brainer date picker. A typical datepicker, say [jQuery UI datepicker](https://jqueryui.com/datepicker/), leaves the model in a date format. I leave it up to you, the owners, to decide whether you'd like to allow for a date string mode.

Without such a mode, I still would like to be able to have something out-of-the-box that supports want I consider the default datepicker case: one that converts date strings to dates and emits date strings. Unfortunately, because the javascript date object is natively formatted in utc, this a problem only a dateInterceptor can solve. Barring a date string mode, I'd like to salvage some ideas here: an out-of-the-box dateInterceptor and convenience date helpers for min-date/max-date comparisons (moving them to dateParser).

wesleycho commented:
UI Bootstrap isn't meant to necessarily be an out of the box library - it is meant to be more of an extensible library that does most of the difficult common operations to avoid a bloating of configuration options (which inevitably leads to configuration options that override the behavior of another configuration option, as well as library bloat). The datepicker has been the antithesis of that unfortunately, and we'd ideally like to keep it as thin as possible.

The description sounds like the opposite of our goals.

davious commented:
OK. Hmm. I sense tension. I'll presume a no on a date string mode.

What are your thoughts on an out-of-the-box dateInterceptor and convenience date helpers for min-date/max-date comparisons (moving them to dateParser)?
